File: spring-aws-s3-java/src/main/java/com/migangqui/spring/aws/s3/service/AmazonS3ServiceImpl.java
package com.migangqui.spring.aws.s3.service;

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.*;
import com.amazonaws.util.IOUtils;
import com.migangqui.spring.aws.s3.bean.UploadFileResult;
import com.migangqui.spring.aws.s3.property.AmazonS3Properties;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.HttpStatus;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;

@Slf4j
@Service
public class AmazonS3ServiceImpl implements AmazonS3Service {

	private AmazonS3 amazonS3Client;

	private AmazonS3Properties properties;

	@Autowired
	public AmazonS3ServiceImpl(AmazonS3 s3Client, AmazonS3Properties properties) {
		this.amazonS3Client = s3Client;
		this.properties = properties;
	}

	@Override
	public UploadFileResult uploadFile(InputStream stream, String folder, String name, String contentType) {
		UploadFileResult result;
		try {
			InputStream streamToUpload = clone(stream);

			ObjectMetadata metadata = new ObjectMetadata();
			metadata.setContentLength(IOUtils.toByteArray(stream).length);

			if (!StringUtils.isEmpty(contentType)) {
				metadata.setContentType(contentType);
				metadata.setCacheControl("s-maxage");
			}

			String path = folder.concat("/").concat(name);

			uploadFileToS3(path, streamToUpload, metadata);

			result = UploadFileResult.builder().fileName(name).status(HttpStatus.SC_OK).build();
		} catch (AmazonServiceException ase) {
			showAmazonServiceExceptionUploadFileLogs(ase);
			result = UploadFileResult.builder().fileName(name).status(HttpStatus.SC_INTERNAL_SERVER_ERROR)
					.cause(ase.getErrorMessage()).exception(ase).build();
		} catch (AmazonClientException ace) {
			showAmazonClientExceptionUploadFileLogs(ace);
			result = UploadFileResult.builder().fileName(name).status(HttpStatus.SC_INTERNAL_SERVER_ERROR)
					.cause(ace.getMessage()).exception(ace).build();
		} catch (Exception e) {
			log.error(e.getMessage(), e);
			result = UploadFileResult.builder().fileName(name).status(HttpStatus.SC_INTERNAL_SERVER_ERROR)
					.cause(e.getMessage()).exception(e).build();
		}
		return result;
	}

	@Override
	public UploadFileResult uploadFile(byte[] bytes, String folder, String name, String contentType) {
		return uploadFile(new ByteArrayInputStream(bytes), folder, name, contentType);
	}
	
	@Async
	@Override
	public UploadFileResult uploadFileAsync(InputStream stream, String folder, String name, String contentType) {
		return uploadFile(stream, folder, name, contentType);
	}

	@Async
	@Override
	public UploadFileResult uploadFileAsync(byte[] bytes, String folder, String name, String contentType) {
		return uploadFileAsync(new ByteArrayInputStream(bytes), folder, name, contentType);
	}

	@Override
	public InputStream getFile(String path) {
		log.info("Reading file from AmazonS3 {}", path);
		S3Object s3Object = amazonS3Client.getObject(new GetObjectRequest(properties.getBucketName(), path));
		return s3Object.getObjectContent();
	}

	@Override
	public boolean deleteFile(String path) {
		log.info("Deleting file from path {}", path);
		boolean result = false;
		try {
			DeleteObjectRequest request = new DeleteObjectRequest(properties.getBucketName(), path);
			amazonS3Client.deleteObject(request);
			result = true;
		} catch (AmazonServiceException ase) {
			showAmazonServiceExceptionUploadFileLogs(ase);
		} catch (AmazonClientException ace) {
			showAmazonClientExceptionUploadFileLogs(ace);
		} catch (Exception e) {
			log.error(e.getMessage(), e);
		}
		return result;
	}

	/* Private methods */

	private void uploadFileToS3(String path, InputStream stream, ObjectMetadata metadata) {
		PutObjectRequest request = new PutObjectRequest(properties.getBucketName(), path, stream, metadata)
				.withCannedAcl(CannedAccessControlList.PublicRead);
		log.debug("Uploading file to {}", path);
		amazonS3Client.putObject(request);
	}

	private void showAmazonServiceExceptionUploadFileLogs(AmazonServiceException ase) {
		log.error("Caught an AmazonServiceException, which means your request made it "
				+ "to Amazon S3, but was rejected with an error response for some reason.");
		log.error("Error Message:    {}", ase.getMessage());
		log.error("HTTP Status Code: {}", ase.getStatusCode());
		log.error("AWS Error Code:   {}", ase.getErrorCode());
		log.error("Error Type:       {}", ase.getErrorType());
		log.error("Request ID:       {}", ase.getRequestId());
	}

	private void showAmazonClientExceptionUploadFileLogs(AmazonClientException ace) {
		log.error("Caught an AmazonClientException, which means the client encountered "
				+ "an internal error while trying to communicate with S3, such as not being able to access the network.");
		log.error("Error Message: " + ace.getMessage());
	}

	private InputStream clone(final InputStream inputStream) {
		InputStream result = null;
		try {
			inputStream.mark(0);
			ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
			byte[] buffer = new byte[1024];
			int readLength = 0;
			while ((readLength = inputStream.read(buffer)) != -1) {
				outputStream.write(buffer, 0, readLength);
			}
			inputStream.reset();
			outputStream.flush();
			result = new ByteArrayInputStream(outputStream.toByteArray());
		}
		catch (Exception ex) {
			ex.printStackTrace();
		}
		return result;
	}
}

File: spring-aws-s3-kotlin/src/main/kotlin/com/migangqui/spring/aws/s3/service/AmazonS3ServiceImpl.kt
package com.migangqui.spring.aws.s3.service

import com.amazonaws.AmazonClientException
import com.amazonaws.AmazonServiceException
import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.model.*
import com.amazonaws.util.IOUtils
import com.migangqui.spring.aws.s3.bean.UploadFileResult
import com.migangqui.spring.aws.s3.property.AmazonS3Properties
import mu.KotlinLogging
import org.apache.http.HttpStatus
import org.springframework.scheduling.annotation.Async
import org.springframework.scheduling.annotation.AsyncResult
import org.springframework.stereotype.Service
import org.springframework.util.StringUtils
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.util.concurrent.Future

@Service
class AmazonS3ServiceImpl(private val s3Client: AmazonS3, private val properties: AmazonS3Properties) : AmazonS3Service {

    private val log = KotlinLogging.logger {}

    override fun uploadFile(stream: InputStream, folder: String, name: String, contentType: String): UploadFileResult {
        var result: UploadFileResult
        try {
            val streamToUpload = stream.clone()

            val metadata = ObjectMetadata()
            metadata.contentLength = IOUtils.toByteArray(stream).size.toLong()

            if (!StringUtils.isEmpty(contentType)) {
                metadata.contentType = contentType
                metadata.cacheControl = "s-maxage"
            }

            val path = "$folder/$name"

            val request = PutObjectRequest(properties.bucketName, path, stream, metadata)
                    .withCannedAcl(CannedAccessControlList.PublicRead)

            log.debug("Uploading file to $path")

            s3Client.putObject(request)

            result = UploadFileResult(name, HttpStatus.SC_OK)
        } catch (ase: AmazonServiceException) {
            showAmazonServiceExceptionUploadFileLogs(ase)
            result = UploadFileResult(name, HttpStatus.SC_INTERNAL_SERVER_ERROR, ase.errorMessage, ase)
        } catch (ace: AmazonClientException) {
            showAmazonClientExceptionUploadFileLogs(ace)
            result = UploadFileResult(name, HttpStatus.SC_INTERNAL_SERVER_ERROR, ace.message, ace)
        } catch (e: Exception) {
            log.error(e.message, e)
            result = UploadFileResult(name, HttpStatus.SC_INTERNAL_SERVER_ERROR, e.message, e)
        }

        return result
    }

    override fun uploadFile(bytes: ByteArray, folder: String, name: String, contentType: String): UploadFileResult {
        return uploadFile(ByteArrayInputStream(bytes), folder, name, contentType)
    }

    @Async
    override fun uploadFileAsync(stream: InputStream, folder: String, name: String, contentType: String): Future<UploadFileResult> {
        return AsyncResult(uploadFile(stream, folder, name, contentType))
    }

    @Async
    override fun uploadFileAsync(bytes: ByteArray, folder: String, name: String, contentType: String): Future<UploadFileResult> {
        return AsyncResult(uploadFile(ByteArrayInputStream(bytes), folder, name, contentType))
    }

    override fun getFile(path: String): InputStream {
        log.info("Reading file from AmazonS3 $path")
        val s3Object = s3Client.getObject(GetObjectRequest(properties.bucketName, path))
        return s3Object.objectContent
    }

    override fun deleteFile(path: String): Boolean {
        log.info("Deleting file from path $path")
        var result = false
        try {
            val request = DeleteObjectRequest(properties.bucketName, path)
            s3Client.deleteObject(request)
            result = true
        } catch (ase: AmazonServiceException) {
            showAmazonServiceExceptionUploadFileLogs(ase)
        } catch (ace: AmazonClientException) {
            showAmazonClientExceptionUploadFileLogs(ace)
        } catch (e: Exception) {
            log.error(e.message, e)
        }
        return result
    }

    /* Private methods */

    private fun showAmazonServiceExceptionUploadFileLogs(ase: AmazonServiceException) {
        log.error("Caught an AmazonServiceException, which means your request made it " +
                "to Amazon S3, but was rejected with an error response for some reason.")
        log.error("Error Message:    ${ase.message}")
        log.error("HTTP Status Code: ${ase.statusCode}")
        log.error("AWS Error Code:   ${ase.errorCode}")
        log.error("Error Type:       ${ase.errorType}")
        log.error("Request ID:       ${ase.requestId}")
    }

    private fun showAmazonClientExceptionUploadFileLogs(ace: AmazonClientException) {
        log.error("Caught an AmazonClientException, which means the client encountered " +
                "an internal error while trying to communicate with S3, such as not being able to access the network.")
        log.error("Error Message: ${ace.message}")
    }

    private fun InputStream.clone(): InputStream? {
        var result: InputStream? = null
        try {
            this.mark(0)
            val outputStream = ByteArrayOutputStream()
            val buffer = ByteArray(1024)
            var readLength = this.read(buffer)
            while ((readLength) != -1) {
                outputStream.write(buffer, 0, readLength)
                readLength = this.read(buffer)
            }
            this.reset()
            outputStream.flush()
            result = ByteArrayInputStream(outputStream.toByteArray())
        } catch (ex: Exception) {
            ex.printStackTrace()
        }

        return result
    }
}